src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp

```diff
@@ -275,6 +278,9 @@ void S3ObjectStorage::findAllFiles(const std::string & path, RelativePathsWithSi
         for (const auto & object : objects)
             children.emplace_back(object.GetKey(), object.GetSize());
 
+        if (max_keys && children.size() >= static_cast<size_t>(max_keys))
+            break;
+
         request.SetContinuationToken(outcome.GetResult().GetNextContinuationToken());
```

Comment:
Should we set `request.SetMaxKeys(max_keys - children.size())` here?
It looks a bit uncomplete strange if we don't.